crates/agglayer-node/src/rpc/mod.rs

```diff
@@ -139,6 +139,12 @@ where
             return Err(Error::rollup_not_registered(tx.tx.rollup_id));
         }
 
+        // Check rate limiting early so that we can report it before entering
+        // the subsequent expensive checks.
+        self.kernel
+            .rate_limiter()
+            .check_send_tx_now(tx.tx.rollup_id)?;
```

Comment:
Does it mean that if 10 tx are proceed at the same time, all of them are proceed for the sanity check but only one are settled and the 9 other got an error ?
Shouldn't it be better to discard the call directly?

Reply:
The question is whether the participating network should be rate limited even if the checks error out and no settlement actually happens. I did not want to apply the settlement rate limit unless there are good chances settlement happens. If that's not the desired behaviour, let me know.

As a side note, I also experimented with a more accurate approach where a rate limiting slot would be "reserved" when the call is first made and then recorded on settlement or released on error. However, it ended up being quite complex so I opted for two checks even though it may lead to spurious errors as you describe.

Reply:
> As we scope the rate limiting per network, we could move the `check_signature` out from the `try_join`, in order to execute it first and then, if it passes, we're sure that that's the right network that submitted this `SignedTx`, thus we can directly apply the rate limiting, as having multiple submission isn't allowed.

I moved the rate limiting check after signature check. However, there is still the question whether a transaction submission that fails the checks should still be counted towards the rate limit even though no settlement actually occurs in that case. There are two checks exactly to avoid that scenario. The first check only checks whether the rate limiter is "full", the other one just before settlement actually records the event in the rate limiter.

Reply:
Indeed, maybe we need to think about that as slot per network?

Reply:
> Indeed, maybe we need to think about that as slot per network?

I am not sure what you mean? The settlement limit is applied on the per-network basis.

Reply:
Rethinking about this, another solution would be to "queue" the queries by fetching an async `lockguard` per network, it would still allow the network to send queries, but only one would be settled.